server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,440 @@
+    }
+
+    private synchronized void enqueueBacklogged() {
+        int maxBackloggedTasksToEnqueue = config.getMaxThreadCount() - currentlyRunningMergeTasks.size();
+        while (maxBackloggedTasksToEnqueue-- > 0) {
+            MergeTask backloggedMergeTask = backloggedMergeTasks.poll();
+            if (backloggedMergeTask == null) {
+                break;
+            }
+            threadPoolMergeQueue.enqueueMergeTask(backloggedMergeTask);
+        }
+    }
+
+    /**
+     * Does the actual merge, by calling {@link org.apache.lucene.index.MergeScheduler.MergeSource#merge}
+     */
+    protected void doMerge(MergeSource mergeSource, MergePolicy.OneMerge merge) throws IOException {
+        mergeSource.merge(merge);
+    }
+
+    @Override
+    public Directory wrapForMerge(MergePolicy.OneMerge merge, Directory in) {
+        // Return a wrapped Directory which has rate-limited output.
+        // Note: the rate limiter is only per thread. So, if there are multiple merge threads running
+        // and throttling is required, each thread will be throttled independently.
+        // The implication of this, is that the total IO rate could be higher than the target rate.
+        RateLimiter rateLimiter = onGoingMergeRateLimiter.get();
+        return new FilterDirectory(in) {
+            @Override
+            public IndexOutput createOutput(String name, IOContext context) throws IOException {
+                ensureOpen();
+
+                // This Directory is only supposed to be used during merging,
+                // so all writes should have MERGE context, else there is a bug
+                // somewhere that is failing to pass down the right IOContext:
+                assert context.context() == IOContext.Context.MERGE : "got context=" + context.context();
+
+                return new RateLimitedIndexOutput(rateLimiter, in.createOutput(name, context));
+            }
+        };
+    }
+
+    final class MergeTask extends AbstractRunnable implements Comparable<MergeTask> {
+        private final String name;
+        private final SetOnce<Long> mergeStartTimeNS;
+        private final MergeSource mergeSource;
+        private final OnGoingMerge onGoingMerge;
+        private final MergeRateLimiter rateLimiter;
+        private final boolean supportsIOThrottling;
+
+        MergeTask(MergeSource mergeSource, MergePolicy.OneMerge merge, boolean supportsIOThrottling, String name) {
+            this.name = name;
+            this.mergeStartTimeNS = new SetOnce<>();
+            this.mergeSource = mergeSource;
+            this.onGoingMerge = new OnGoingMerge(merge);
+            this.rateLimiter = new MergeRateLimiter(merge.getMergeProgress());
+            this.supportsIOThrottling = supportsIOThrottling;
+            // probably redundant, but better be explicit
+            if (this.supportsIOThrottling == false) {
+                this.rateLimiter.setMBPerSec(Double.POSITIVE_INFINITY);
+            }
```

Comment:
This is the default and we expect something to later set a reduced rate if necessary. I'd probably prefer not having this just like the existing `MergeThread` does not.

Reply:
> I'd probably prefer not having this just like the existing MergeThread does not.

I agreee. Pushed a315198